fem/libceed/ceed.cpp

```diff
@@ -60,8 +62,234 @@ void InitCeedCoeff(Coefficient* Q, CeedData* ptr)
    {
       MFEM_ABORT("This type of Coefficient is not supported.");
    }
+#else
+   mfem_error("MFEM must be built with MFEM_USE_CEED=YES to use libCEED.");
+#endif
+}
+
+void CeedPAAssemble(const CeedPAOperator& op,
+                    CeedData& ceedData)
+{
+#ifdef MFEM_USE_CEED
+   const FiniteElementSpace &fes = op.fes;
+   const mfem::IntegrationRule &irm = op.ir;
+   Ceed ceed(internal::ceed);
+   mfem::Mesh *mesh = fes.GetMesh();
+   CeedInt nqpts, nelem = mesh->GetNE();
+   CeedInt dim = mesh->SpaceDimension(), vdim = fes.GetVDim();
+
+   mesh->EnsureNodes();
+   InitCeedBasisAndRestriction(fes, irm, ceed, &ceedData.basis, &ceedData.restr);
+
+   const mfem::FiniteElementSpace *mesh_fes = mesh->GetNodalFESpace();
+   MFEM_VERIFY(mesh_fes, "the Mesh has no nodal FE space");
+   InitCeedBasisAndRestriction(*mesh_fes, irm, ceed, &ceedData.mesh_basis,
+                               &ceedData.mesh_restr);
+
+   CeedBasisGetNumQuadraturePoints(ceedData.basis, &nqpts);
+
+   const int qdatasize = op.qdatasize;
+   CeedElemRestrictionCreateStrided(ceed, nelem, nqpts, qdatasize,
+                                    nelem*nqpts*qdatasize, CEED_STRIDES_BACKEND,
+                                    &ceedData.restr_i);
+
+   CeedVectorCreate(ceed, mesh->GetNodes()->Size(), &ceedData.node_coords);
+   CeedScalar *nodes_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(ceed, &mem);
+   if ( Device::Allows(Backend::DEVICE_MASK) && mem==CEED_MEM_DEVICE )
+   {
+      nodes_ptr = const_cast<CeedScalar*>(mesh->GetNodes()->Read());
+   }
+   else
+   {
+      nodes_ptr = const_cast<CeedScalar*>(mesh->GetNodes()->HostRead());
+      mem = CEED_MEM_HOST;
+   }
+   CeedVectorSetArray(ceedData.node_coords, mem, CEED_USE_POINTER, nodes_ptr);
```

Comment:
This logic for wrapping an `mfem::Vector` as `CeedVector` is used in a number of places, so we should probably turn these into functions to reduce duplication.

Reply:
This is a good idea.

Reply:
I've done it directly in #1719 , there was even more code duplication there.